File: app/src/main/java/com/i14yokoro/mldpterminal/HtmlParser.java
package com.i14yokoro.mldpterminal;

import android.graphics.Typeface;
import android.text.Spannable;
import android.text.SpannableStringBuilder;
import android.text.style.StyleSpan;
import android.text.style.UnderlineSpan;

public class HtmlParser {
    public static String toHtml(Spannable text) {
        final SpannableStringBuilder ssBuilder = new SpannableStringBuilder(text);
        int start, end;

        StyleSpan[] styleSpans = ssBuilder.getSpans(0, text.length(), StyleSpan.class);
        for (int i = styleSpans.length - 1; i >= 0; i--) {
            StyleSpan span = styleSpans[i];
            start = ssBuilder.getSpanStart(span);
            end = ssBuilder.getSpanEnd(span);
            ssBuilder.removeSpan(span);
            if (span.getStyle() == Typeface.BOLD) {
                ssBuilder.insert(start, "<b>");
                ssBuilder.insert(end + 3, "</b>");
            } else if (span.getStyle() == Typeface.ITALIC) {
                ssBuilder.insert(start, "<i>");
                ssBuilder.insert(end + 3, "</i>");
            }
        }

        UnderlineSpan[] underSpans = ssBuilder.getSpans(0, ssBuilder.length(), UnderlineSpan.class);
        for (int i = underSpans.length - 1; i >= 0; i--) {
            UnderlineSpan span = underSpans[i];
            start = ssBuilder.getSpanStart(span);
            end = ssBuilder.getSpanEnd(span);
            ssBuilder.removeSpan(span);
            ssBuilder.insert(start, "<u>");
            ssBuilder.insert(end + 3, "</u>");
        }
        replace(ssBuilder, '\n', "<br/>");

        return ssBuilder.toString();
    }

    private static void replace(SpannableStringBuilder b, char oldChar, String newStr) {
        for (int i = b.length() - 1; i >= 0; i--) {
            if (b.charAt(i) == oldChar) {
                b.replace(i, i + 1, newStr);
            }
        }
    }
}

File: app/src/main/kotlin/com/i14yokoro/mldpterminal/HtmlParser.kt
package com.i14yokoro.mldpterminal

import android.graphics.Typeface
import android.text.Spannable
import android.text.SpannableStringBuilder
import android.text.style.StyleSpan
import android.text.style.UnderlineSpan

object HtmlParser {
    fun toHtml(text: Spannable): String {
        val ssBuilder = SpannableStringBuilder(text)
        var start: Int
        var end: Int

        val styleSpans = ssBuilder.getSpans<StyleSpan>(0, text.length, StyleSpan::class.java)
        for (i in styleSpans.indices.reversed()) {
            val span = styleSpans[i]
            start = ssBuilder.getSpanStart(span)
            end = ssBuilder.getSpanEnd(span)
            ssBuilder.removeSpan(span)
            if (span.style == Typeface.BOLD) {
                ssBuilder.insert(start, "<b>")
                ssBuilder.insert(end + 3, "</b>")
            } else if (span.style == Typeface.ITALIC) {
                ssBuilder.insert(start, "<i>")
                ssBuilder.insert(end + 3, "</i>")
            }
        }

        val underSpans = ssBuilder.getSpans<UnderlineSpan>(0, ssBuilder.length, UnderlineSpan::class.java)
        for (i in underSpans.indices.reversed()) {
            val span = underSpans[i]
            start = ssBuilder.getSpanStart(span)
            end = ssBuilder.getSpanEnd(span)
            ssBuilder.removeSpan(span)
            ssBuilder.insert(start, "<u>")
            ssBuilder.insert(end + 3, "</u>")
        }
        replace(ssBuilder, '\n', "<br/>")

        return ssBuilder.toString()
    }

    private fun replace(b: SpannableStringBuilder, oldChar: Char, newStr: String) {
        for (i in b.length - 1 downTo 0) {
            if (b[i] == oldChar) {
                b.replace(i, i + 1, newStr)
            }
        }
    }
}